### PR TITLE
Created request_header_provider and implemented databricks_request_header

### PR DESCRIPTION
**Pull request by jimmyxu-db**
## What changes are proposed in this pull request?

Adding a new plugin interface: the RequestHeaderProvider. This plugin will create a dict of request headers to attach to outgoing requests. The actual attaching to requests will happen in a follow up PR.

Implemented the DatabricksRequestHeaderProvider, which creates request headers for databricks environments.

## How is this patch tested?

Unit tests

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [x] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/request_header/databricks_request_header_provider.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+
+class DatabricksRequestHeaderProvider(RequestHeaderProvider):
+    def in_context(self):
+        return databricks_utils.is_in_cluster()
+
+    def request_headers(self):
+        request_headers = {}
+        if databricks_utils.is_in_databricks_notebook():
+            print("hello")
```
Comment (mohamad-arabi) [inline mlflow/tracking/request_header/databricks_request_header_provider.py:12]: Do you want to remove this?

File: mlflow/tracking/request_header/databricks_request_header_provider.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+            request_headers["notebook_id"] = databricks_utils.get_notebook_id()
+        if databricks_utils.is_in_databricks_job():
```
Comment (mohamad-arabi) [inline mlflow/tracking/request_header/databricks_request_header_provider.py:14]: Are these if statements mutually exclusive? should these be `if` or `else if`. Can `is_in_databricks_notebook ` and `is_in_databricks_job ` both be true?

File: mlflow/tracking/request_header/databricks_request_header_provider.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+from mlflow.tracking.request_header.abstract_request_header_provider import RequestHeaderProvider
+from mlflow.utils import databricks_utils
+
+
+class DatabricksRequestHeaderProvider(RequestHeaderProvider):
+    def in_context(self):
+        return databricks_utils.is_in_cluster()
```
Comment (dbczumar) [inline mlflow/tracking/request_header/databricks_request_header_provider.py:7]: Is the cluster ID guaranteed to be present in all contexts? I'm wondering if instead we should return `is_in_cluster() or is_in_databricks_notebook() or is_in_databricks_job()` to make sure that we still use a subset of headers even if the cluster ID can't be fetched

File: tests/tracking/request_header/test_databricks_request_header_provider.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+from mlflow.tracking.request_header.databricks_request_header_provider import (
+    DatabricksRequestHeaderProvider,
+)
+
+
+def test_databricks_cluster_run_context_in_context():
+    with mock.patch("mlflow.utils.databricks_utils.is_in_cluster") as in_cluster_mock:
+        assert DatabricksRequestHeaderProvider().in_context() == in_cluster_mock.return_value
+
+
```
Comment (dbczumar) [inline tests/tracking/request_header/test_databricks_request_header_provider.py:13]: Can we add a couple of tests for the case where we're in multiple contexts, ensuring that all relevant information is captured: one for the case where we have cluster ID + notebook info available and one for the case where we have cluster ID + job info available?

File: mlflow/tracking/request_header/databricks_request_header_provider.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/tracking/request_header/databricks_request_header_provider.py:14]: While it's probably not the case that notebook id / job id can be present together, the MLflow client shouldn't consider it an error case if they are. Notebook id can be present with cluster id, and job id can also be present with cluster id.

Accordingly, I think it's best to capture all information that's available, preferring the repeated `if` structure over `if, elif`.

File: mlflow/tracking/request_header/databricks_request_header_provider.py (hunk 3 above)
Comment (jimmyxu-db) [reply mlflow/tracking/request_header/databricks_request_header_provider.py:7]: Good idea!

File: mlflow/tracking/request_header/databricks_request_header_provider.py (hunk 2 above)
Comment (jimmyxu-db) [reply mlflow/tracking/request_header/databricks_request_header_provider.py:14]: i think it makes sense to do separate ifs to return what we have

File: tests/tracking/request_header/test_databricks_request_header_provider.py (hunk 4 above)
Comment (jimmyxu-db) [reply tests/tracking/request_header/test_databricks_request_header_provider.py:13]: Do you mean to update the `in_context()` to be an OR of the 3 environments and then update this test accordingly?

Comment (jimmyxu-db) [reply tests/tracking/request_header/test_databricks_request_header_provider.py:13]: Don't the existing tests add whatever info we have available?

Comment (dbczumar) [reply tests/tracking/request_header/test_databricks_request_header_provider.py:32]: They don’t verify that we add cluster id and notebook info if both are available. I.e. I’m pretty sure that all tests would pass if we used the kind of if / elif logic that mohamad asked about above.

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @jimmyxu-db !

**Pull request merged**